src/java/org/apache/cassandra/index/sai/QueryContext.java

```diff
@@ -66,6 +69,8 @@
 
     private final Map<SSTableReader, SSTableQueryContext> sstableQueryContexts = new HashMap<>();
 
+    public final ConcurrentSkipListSet<PrimaryKey> shadowedPrimaryKeys = new ConcurrentSkipListSet<>();
```

Comment:
does this need to be Concurrent or will a HashSet be fine?

Reply:
It doesn't have to be concurrent, but good to be sorted, so it can abort early if exceed segment boundary

Reply:
ah, I will make it a TreeSet and make the return type of the getter a Navigable so that is clear.  thanks